pkg/logql/log/parser_hints.go

```diff
@@ -25,26 +25,41 @@ type ParserHint interface {
 	//		 sum(rate({app="foo"} | json [5m]))
 	// We don't need to extract any labels from the log line.
 	NoLabels() bool
+
+	// Holds state about what's already been extracted for the associated
+	// labels. This assumes that only required labels are ever extracted
+	RecordExtracted(string)
+	AllRequiredExtracted() bool
+	Reset()
 }
 
-type parserHint struct {
+type Hints struct {
 	noLabels       bool
 	requiredLabels []string
+	extracted      []string
 }
 
-func (p *parserHint) ShouldExtract(key string) bool {
+func (p *Hints) ShouldExtract(key string) bool {
 	if len(p.requiredLabels) == 0 {
 		return true
 	}
+
+	for _, l := range p.extracted {
```

Comment:
You mean the already extracted labels, right?

Reply:
yes